Create eslint-plugin-eui-a11y plugin

CoenWarmer opened the pull request:
## Summary

This adds an additional custom ESLint rule package which checks certain Eui elements for the existence of an `aria-label` prop.

If it exists, it will leave it untouched. If it doesn't, it will warn the engineer it needs to be added, and offers a autofix suggestion for those engineers who have fix on save enabled in their IDE.

<img width="739" alt="Screenshot 2025-03-25 at 13 59 28" src="https://github.com/user-attachments/assets/0813b317-c752-40d7-b569-e866a3ecf6b0" />

<img width="804" alt="Screenshot 2025-03-25 at 13 59 36" src="https://github.com/user-attachments/assets/3c45c49c-6db8-4740-b5de-89aa534c248b" />

This package is an offshoot of the `kbn-eslint-plugin-i18n` and `kbn-eslint-plugin-telemetry` packages.

CoenWarmer committed:
Create eslint-plugin-eui-a11y plugin

CoenWarmer committed:
Enable eui-a11y rule for APM, Infra and Shared

CoenWarmer committed:
Merge branch 'main' of github.com:elastic/kibana into feat/create-eui-a11y-eslint-plugin

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

jennypavlova commented on packages/kbn-eslint-plugin-eui-a11y/helpers/get_intent_from_node.ts:
```diff
@@ -0,0 +1,126 @@
+
+      case 'JSXElement':
+        // Determining whether node is of form `<FormattedMessage defaultMessage="message" />`
+        const name: TSESTree.JSXTagNameExpression = currentNode.openingElement.name;
```
NIT should we us a different name like `openingElement` ot `jsxTagObject` or something else so we won't have the `name.name` on[ line 48](https://github.com/elastic/kibana/pull/215859/files#diff-5137898a4b652d250758238bea6b84edb4e4ead45bb64aa86b10bdb40fc5160cR48)

CoenWarmer replied:
Agreed. Will adjust.

jennypavlova commented on packages/kbn-eslint-plugin-eui-a11y/rules/eui_elements_should_have_a11y_props.ts:
```diff
@@ -0,0 +1,84 @@
+          sourceCode.getScope(node as TSNode)
+        );
+
+        if (hasDataTestSubjProp) {
```
Is this a valid case? Do we want to skip it if we have `data-test-subj` or am I missing something 🤔

CoenWarmer replied:
Variable name was left unadjusted when duplicating the existing ESLint custom rule package. Will double check the condition.

jennypavlova commented on packages/kbn-eslint-plugin-eui-a11y/rules/eui_elements_should_have_a11y_props.ts:
```diff
@@ -0,0 +1,84 @@
+            .match(/[A-Z][a-z]*/g) || []
+        ).join(' ');
+
+        const suggestion = `${intent}${element}`; // 'Actions Button'
```
NIT: Should we translate the suggestion to avoid fixing another rule after?

jennypavlova replied:
Example: 
<img width="1066" alt="image" src="https://github.com/user-attachments/assets/38ae227e-16ff-4987-8b9a-d6407348d552" />

CoenWarmer replied:
Is a really good point! The aria-label of course also needs to be i18n-ified.

The logic of this rule might need to be incorporated inside the existing i18n ESLint custom rule. Will pick this up.

jennypavlova reviewed (commented):
Nice, thanks for adding that! Added some questions/nits ⬇️

iblancof commented on packages/kbn-eslint-plugin-eui-a11y/rules/eui_elements_should_have_a11y_props.ts:
```diff
@@ -0,0 +1,84 @@
+import { checkNodeForExistingAriaLabelProp } from '../helpers/check_node_for_existing_data_test_subj_prop';
+import { getIntentFromNode } from '../helpers/get_intent_from_node';
+
+export const EUI_ELEMENTS = [
```
Where is this list from? Are we sure there aren’t any other elements that need the aria-label?

iblancof replied:
Should we start a conversation with the EUI team about this? 
I'm not suggesting blocking the PR, just looking to get the ball rolling on having a consistent strategy for a11y across Kibana.

weronikaolejniczak replied:
@iblancof @CoenWarmer hey, EUI team member here! 👋🏻 While an ESLint plugin can be used to handle these cases, I believe an even simpler solution is to use appropriate typings for components (i.e. require `label` or `aria-label`, depending on the case). That would also allow us to control the a11y best practices from EUI and would rid you of the maintenance burden.

In any case, this initiative is awesome and it'd be amazing to discuss it together 😄

iblancof commented on packages/kbn-eslint-plugin-eui-a11y/rules/eui_elements_should_have_a11y_props.ts:
```diff
@@ -0,0 +1,84 @@
+  'EuiSelectWithWidth',
+];
+
+export const EuiElementsShouldHaveA11yProps: Rule.RuleModule = {
```
The naming seems quite generic and not specifically tied to the aria-label attribute. Do we plan to add more tags to it?

jennypavlova replied:
As a follow-up here - do we plan to extend this to support also `role`, for example, or other properties? If not, maybe changing the name will make sense

jbudz reviewed (approved):
`packages/kbn-eslint-config/.eslintrc.js` LGTM

MiriamAparicio commented on .eslintrc.js:
```diff
@@ -1010,6 +1010,16 @@ module.exports = {
+      files: [
+        'x-pack/solutions/observability/plugins/apm/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
+        'x-pack/solutions/observability/plugins/infra/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
+        'x-pack/solutions/observability/plugins/shared/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
```
We might want to add profiling here as well

miloszmarcinkowski commented on .eslintrc.js:
```diff
@@ -1010,6 +1010,16 @@ module.exports = {
+      files: [
+        'x-pack/solutions/observability/plugins/apm/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
+        'x-pack/solutions/observability/plugins/infra/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
+        'x-pack/solutions/observability/plugins/shared/**/!(*.stories.tsx|*.test.tsx|*.storybook_decorator.tsx|*.mock.tsx)',
```
Does `x-pack/solutions/observability/plugins/shared` path exist?

I applied [those](https://github.com/elastic/kibana/blob/main/.eslintrc.js#L1026) paths when I was adding eslint rule for our code ownership. Though, I didn't verify it, just looked into CODEOWNERS file.

CoenWarmer replied:
Nice, it's actually `observability_shared`.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/286933)
* Commit: 0601da3c5e781b3541115bce1f1b078db316525c

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286933","number":286933,"commit":"0601da3c5e781b3541115bce1f1b078db316525c"}],"number":286933}
buildkite-pr-comment-->

alexwizp commented:
@CoenWarmer I'm not sure we need to push developers to add an `aria-label` in every case. Sometimes it's unnecessary, for example, when using `EuiFormRow` or similar components.

Additionally, we can sometimes use `aria-labelledby` instead. These are just two cases where your approach doesn't seem ideal.

bhavyarm commented:
@CoenWarmer Thanks for doing this work. I am new to this and had a question for you. Could you please give me some examples of what would be the result of autofix? Would it suggest generic aria-labels? Thanks!

CoenWarmer commented:
@alexwizp this doesn't add the prop to every Eui element. Just a subset where it always makes sense.

alexwizp commented:
Just wanted to say that your rule should correctly works with following cases

1. No need to pass `aria-label` in case of using inside of EiuFormRow
```
<EuiFormRow label="Text field" helpText="I am some friendly help text.">
    <EuiFieldText name="first" />
  </EuiFormRow>
```

2. No need to pass  `aria-label` in case of using `aria-labelledby`

```
<label id="someId">Label</label>
<EuiFieldText name="first" aria-labelledby="someId"/>
```

CoenWarmer committed:
Addressed feedback

CoenWarmer committed:
Merge branch 'feat/create-eui-a11y-eslint-plugin' of github.com:CoenWarmer/kibana into feat/create-eui-a11y-eslint-plugin

CoenWarmer committed:
Add translation logic to aria-label prop

alexwizp commented:
@CoenWarmer Thanks! If your code doesn’t trigger any lint issues in code like `examples/field_formats_example/public/examples/3_creating_custom_format_editor.tsx` (EuiFormRow -> EuiSelect), I’m happy with these changes.

Just to clarify, my motivation is simple—I understand that these rules were introduced primarily for `OBLT` code, but I’d prefer to avoid adding extra translations unless they are truly necessary.

jbudz commented on packages/kbn-eslint-plugin-eui-a11y/rules/eui_elements_should_have_aria_label_or_aria_labelledby_props.ts:
```diff
@@ -0,0 +1,113 @@
+import { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/typescript-estree';
+import { Node } from 'estree';
+
+import { checkNodeForExistingAriaLabelProp } from '../helpers/check_node_for_existing_data_test_subj_prop';
```
Should this filename be `check_node_for_existing_aria_label_prop`?

CoenWarmer replied:
Absolutely right! Thanks :)

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/287041)
* Commit: d9dec7c8605a7e68984ee9a30da555ea972992d7

### Failed CI Steps
* [FTR Configs #27](https://buildkite.com/elastic/kibana-pull-request/builds/287041#0195ce4c-32ba-48ec-a519-d0baca9bea87)
* [FTR Configs #64](https://buildkite.com/elastic/kibana-pull-request/builds/287041#0195ce4c-32cd-4772-b180-e62e6e52b355)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/287041#0195ce4c-32cd-4772-b180-e62e6e52b355) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/287041/jobs/0195ce4c-32cd-4772-b180-e62e6e52b355/artifacts/0195ce71-ede9-42b0-bc0d-a454de8de839) FTR Configs #<span></span>64 / EQL execution logic API @<span></span>ess @<span></span>serverless @<span></span>serverlessQA EQL type rules parses shard failures for EQL event query

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #286933](https://buildkite.com/elastic/kibana-pull-request/builds/286933) succeeded 0601da3c5e781b3541115bce1f1b078db316525c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287041","number":287041,"commit":"d9dec7c8605a7e68984ee9a30da555ea972992d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286933","number":286933,"commit":"0601da3c5e781b3541115bce1f1b078db316525c"}],"number":287041}
buildkite-pr-comment-->

rmyz reviewed (approved):
LGTM! Good job 🚀

CoenWarmer committed:
Cleanup

CoenWarmer committed:
Update README

CoenWarmer committed:
Add lowercased element name to i18n translation id

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/287251)
* Commit: 8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/287251#0195d217-a486-418a-b29d-b755dd26eefd)

### History
* :yellow_heart: [Build #287041](https://buildkite.com/elastic/kibana-pull-request/builds/287041) was flaky d9dec7c8605a7e68984ee9a30da555ea972992d7
* :green_heart: [Build #286933](https://buildkite.com/elastic/kibana-pull-request/builds/286933) succeeded 0601da3c5e781b3541115bce1f1b078db316525c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287251","number":287251,"commit":"8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287041","number":287041,"commit":"d9dec7c8605a7e68984ee9a30da555ea972992d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286933","number":286933,"commit":"0601da3c5e781b3541115bce1f1b078db316525c"}],"number":287251}
buildkite-pr-comment-->

CoenWarmer committed:
Add iconType value to intent for EuiButtonIcon

alexwizp commented on packages/kbn-eslint-plugin-eui-a11y/README.mdx:
````diff
@@ -0,0 +1,89 @@
+```
+export const InfraBetaBadge = ({ iconType, tooltipPosition, tooltipContent }: Props) => (
+  <EuiBetaBadge
+    aria-label={i18n.translate('xpack.infra.infraBetaBadge.betaBadge.ariaLabel', {
````
@CoenWarmer sorry for my 2 cents, but why we want to add `aria-label` for EuiBetaBadge? I don't see any a11y issues in case of using only `label`

Can I ask you to discuss your PR with someone from EUI before merge please.  

@bhavyarm fyi

CoenWarmer replied:
These were created here https://github.com/elastic/kibana/issues/212264.

alexwizp replied:
@CoenWarmer it's not explain case which you added as an example

```
<EuiBetaBadge
    label={i18n.translate('xpack.infra.common.tabBetaBadgeLabel', {
      defaultMessage: 'Beta',
    })}
    tooltipContent={tooltipContent}
    iconType={iconType}
    tooltipPosition={tooltipPosition}
    data-test-id="infra-beta-badge"
  />
```
`aria-label` definitely not required here

CoenWarmer committed:
Cleanup

CoenWarmer commented:
> @CoenWarmer Thanks for doing this work. I am new to this and had a question for you. Could you please give me some examples of what would be the result of autofix? Would it suggest generic aria-labels? Thanks!

Of course!

I've added some examples in the README.md, but here they are:

```
/* file is in `x-pack/solutions/observability/plugins/apm/public/components/fleet_integration/apm_agents/runtime_attachment/discovery_rule.tsx` */

<EuiButtonIcon
    iconType="pencil"
    onClick={() => {
        onEdit(id);
    }}
/>
```

turns into:

```
<EuiButtonIcon
    aria-label={i18n.translate('xpack.apm.discoveryRule.pencilButton.ariaLabel', {
        defaultMessage: 'Pencil Button',
    })}
    iconType="pencil"
    onClick={() => {
        onEdit(id);
    }}
/>
```

Another example:

```
/* file is in: `x-pack/solutions/observability/plugins/infra/public/components/beta_badge.tsx` */

export const InfraBetaBadge = ({ iconType, tooltipPosition, tooltipContent }: Props) => (
  <EuiBetaBadge
    label={i18n.translate('xpack.infra.common.tabBetaBadgeLabel', {
      defaultMessage: 'Beta',
    })}
    tooltipContent={tooltipContent}
    iconType={iconType}
    tooltipPosition={tooltipPosition}
    data-test-id="infra-beta-badge"
  />
);
```

gets transformed into:

```
export const InfraBetaBadge = ({ iconType, tooltipPosition, tooltipContent }: Props) => (
  <EuiBetaBadge
    aria-label={i18n.translate('xpack.infra.infraBetaBadge.betaBadge.ariaLabel', {
      defaultMessage: 'Beta Badge',
    })}
    label={i18n.translate('xpack.infra.common.tabBetaBadgeLabel', {
      defaultMessage: 'Beta',
    })}
    tooltipContent={tooltipContent}
    iconType={iconType}
    tooltipPosition={tooltipPosition}
    data-test-id="infra-beta-badge"
  />
);
```

Hope this helps!

CoenWarmer committed:
Merge branch 'main' into feat/create-eui-a11y-eslint-plugin

alexwizp commented:
@elastic/eui-team  could you please review that PR as well? I think forcing the aria-label attribute directly isn’t the best approach. Right now, it only affects the oblt code, but we should find a better way to handle it.

alexwizp commented:
one more point about 

```
<EuiButtonIcon
    aria-label={i18n.translate('xpack.apm.discoveryRule.pencilButton.ariaLabel', {
        defaultMessage: 'Pencil Button',
    })}
    iconType="pencil"
    onClick={() => {
        onEdit(id);
    }}
/> 
```

In screen readers this results in duplicate pronunciation of the word `Button`, e.g. https://github.com/elastic/kibana/issues/215647 .

I understand the desire to merge this PR, but on the other hand I see many edge cases that are not taken into account

CoenWarmer commented:
> one more point about 
> 
> 
> 
> ```
> 
> <EuiButtonIcon
> 
>     aria-label={i18n.translate('xpack.apm.discoveryRule.pencilButton.ariaLabel', {
> 
>         defaultMessage: 'Pencil Button',
> 
>     })}
> 
>     iconType="pencil"
> 
>     onClick={() => {
> 
>         onEdit(id);
> 
>     }}
> 
> /> 
> 
> ```
> 
> 
> 
> In screen readers this results in duplicate pronunciation of the word `Button`, e.g. .
> 
> 

Can you explain how this would happen? 

> 
> I understand the desire to merge this PR, but on the other hand I see many edge cases that are not taken into account

Please note all the edge cases you see so I can try to accommodate for them.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/287433)
* Commit: ed6c27b1a39b8f03fb9643ecd54cbd2264951286

### Failed CI Steps
* [Investigations - Security Solution Cypress Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/287433#0195d357-2cbe-4d98-b476-f6543b6b92c1)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/287433#0195d357-2cbe-4d98-b476-f6543b6b92c1) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/287433/jobs/0195d357-2cbe-4d98-b476-f6543b6b92c1/artifacts/0195d399-0a45-4188-92ef-d5719289d352) Investigations - Security Solution Cypress Tests #<span></span>7 / Detections : Page Filters with data modification should update alert status list when the alerts are updated should update alert status list when the alerts are updated

### History
* :broken_heart: [Build #287251](https://buildkite.com/elastic/kibana-pull-request/builds/287251) failed 8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef
* :yellow_heart: [Build #287041](https://buildkite.com/elastic/kibana-pull-request/builds/287041) was flaky d9dec7c8605a7e68984ee9a30da555ea972992d7
* :green_heart: [Build #286933](https://buildkite.com/elastic/kibana-pull-request/builds/286933) succeeded 0601da3c5e781b3541115bce1f1b078db316525c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287433","number":287433,"commit":"ed6c27b1a39b8f03fb9643ecd54cbd2264951286"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287251","number":287251,"commit":"8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287041","number":287041,"commit":"d9dec7c8605a7e68984ee9a30da555ea972992d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286933","number":286933,"commit":"0601da3c5e781b3541115bce1f1b078db316525c"}],"number":287433}
buildkite-pr-comment-->

alexwizp commented:
But my main concern here

 ```
<EuiFormRow label="Currency">
    <EuiSelect
      defaultValue={1}
      options={[
        { text: 'EUR', value: 'EUR' },
        { text: 'USD', value: 'USD' },
      ]}
      onChange={(e) => {
      }}
    />
</EuiFormRow>
```
For that example, there's no need to force the use of `aria-label `for `EuiSelect`. This applies to any first child of `EuiFormRow`.

CoenWarmer committed:
Remove element type from defaultMessage

CoenWarmer commented:
> @CoenWarmer Thanks for doing this work. I am new to this and had a question for you. Could you please give me some examples of what would be the result of autofix? Would it suggest generic aria-labels? Thanks!

Thanks!

I've just added an example here:

```
/* file is in `x-pack/solutions/observability/plugins/apm/public/components/fleet_integration/apm_agents/runtime_attachment/discovery_rule.tsx` */

<EuiButtonIcon
    data-test-subj="apmDiscoveryRuleButton"
    iconType="pencil"
    color="primary"
    onClick={() => {
        onEdit(id);
    }}
/>
```

turns into:

```
<EuiButtonIcon
    aria-label={i18n.translate('xpack.apm.discoveryRule.button.ariaLabel', {
        defaultMessage: 'Button',
    })}
    data-test-subj="apmDiscoveryRuleButton"
    iconType="pencil"
    color="primary"
    onClick={() => {
        onEdit(id);
    }}
/>
```

> Screen readers typically announce the ARIA role of a DOM element, either before or after the associated label, depending on the specific screen reader. In the example below:
> 
> <img alt="image" width="370" src="https://private-user-images.githubusercontent.com/20072247/427159968-55fa3243-4130-4009-b54d-1df501a46fe4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMTMwMjgsIm5iZiI6MTc0MzAxMjcyOCwicGF0aCI6Ii8yMDA3MjI0Ny80MjcxNTk5NjgtNTVmYTMyNDMtNDEzMC00MDA5LWI1NGQtMWRmNTAxYTQ2ZmU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI2VDE4MTIwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhN2QwNDgyYWRjNzRlYmZjZWM3MTQzNjk2MWU5ZmVkYTgyMTExZjAyMjc1Nzg1MWJiMGEwNjE3YWQ2YTFjZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qYEJV1v1mRgzvOR3ybtSpiE4vD1MUkaBeYgkjapZmKE">
> Users hear the word "button" twice (one from ARIA role, one from label), which is redundant and unnecessary.

This is now fixed.

CoenWarmer commented:
@alexwizp if you can provide a list of different scenarios and in which scenarios elements need or don't need to have an aria-label, I will accomodate for them.

weronikaolejniczak commented:
@alexwizp thank you for bringing awareness to this PR! 😄 

@CoenWarmer this plugin is great, awesome job! 💪🏻  I'm all for a11y improvements. To highlight what @alexwizp wrote, in case the consumer passes a label as a prop or through children, it's not necessary to pass `aria-label`. I have some doubts regarding the auto-fix. The purpose of the label is to inform the user about not only what the element is but also what the element does. I would argue that applying an `aria-label` of "Pencil button" etc. is redundant. The label should be e.g. "Edit user information" or "Save changes" which comes from the usage context, therefore it's not easy to infer and apply in an auto-fix.

One thing that is worth noting as well is that we've recently [improved our ESLint plugin](https://github.com/elastic/eui/pull/8304). At the moment of writing this, I'm releasing the latest changes, and updating the dependency in Kibana and Cloud. I would advise against EUI-specific ESLint plugins in Kibana, especially those that a broader audience can benefit from. Anything that has "EUI" in the name or applies to EUI component sounds like a candidate to add to our existing plugin. Would you consider contributing to [eui](https://github.com/elastic/eui) repo instead?

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/287469)
* Commit: c767c2323a032f03f18858ce191192c8916fa90b

### Failed CI Steps
* [FTR Configs #123](https://buildkite.com/elastic/kibana-pull-request/builds/287469#0195d3c6-7190-4bc4-a0b9-76671a4a6265)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #287433](https://buildkite.com/elastic/kibana-pull-request/builds/287433) failed ed6c27b1a39b8f03fb9643ecd54cbd2264951286
* :broken_heart: [Build #287251](https://buildkite.com/elastic/kibana-pull-request/builds/287251) failed 8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef
* :yellow_heart: [Build #287041](https://buildkite.com/elastic/kibana-pull-request/builds/287041) was flaky d9dec7c8605a7e68984ee9a30da555ea972992d7
* :green_heart: [Build #286933](https://buildkite.com/elastic/kibana-pull-request/builds/286933) succeeded 0601da3c5e781b3541115bce1f1b078db316525c

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287469","number":287469,"commit":"c767c2323a032f03f18858ce191192c8916fa90b"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287433","number":287433,"commit":"ed6c27b1a39b8f03fb9643ecd54cbd2264951286"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287251","number":287251,"commit":"8ec1a50497ac010938dd56b4d3b2f4f5e7a02cef"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/287041","number":287041,"commit":"d9dec7c8605a7e68984ee9a30da555ea972992d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/286933","number":286933,"commit":"0601da3c5e781b3541115bce1f1b078db316525c"}],"number":287469}
buildkite-pr-comment-->

JasonStoltz commented:
@alexwizp Were your concerns addressed on this PR? I think it's important that we get the rules correct here.

alexwizp commented:
@JasonStoltz  I wouldn’t say that much has changed; the rules are still quite straightforward. I also agree with @weronikaolejniczak that these rules should be reviewed and then added to the EUI repository.
